File: src/main/java/lec12/Lec12Main.java
package lec12;

public class Lec12Main {

    public static void main(String[] args) {
        Person.Factory.newBaby("ABC");

        // @JvmStatic 이 있는 경우 이름 또는 Companion 호출 없이 바로 사용 가능
        Person.newBaby("ABC");

        moveSomething(new Movable() {
            @Override
            public void move() {
                System.out.println("움직인다~~~");
            }

            @Override
            public void fly() {
                System.out.println("난다~~~");
            }
        });
    }

    private static void moveSomething(Movable movable) {
        movable.move();
        movable.fly();
    }
}

File: src/main/kotlin/lec12/Lec12Main.kt
package lec12

fun main() {
    println(Singleton.a)
    Singleton.a += 10
    println(Singleton.a)

    moveSomething(object : Movable {
        override fun move() {
            println("움직인다~~~!")
        }

        override fun fly() {
            println("난다~~~!")
        }
    })
}

class Person private constructor(
    var name: String,
    var age: Int
) {
    companion object Factory : Log {
        private const val MIN_AGE = 1

        @JvmStatic
        fun newBaby(name: String): Person {
            return Person(name, MIN_AGE )
        }

        override fun Log() {
            println("나는 Person 클래스의 동행 객체에요")
        }
    }
}

object Singleton {
    var a : Int = 0
}

private fun moveSomething(movable: Movable) {
    movable.move()
    movable.fly()
}